File: ga/gpt3_5_result_2024_6/result/multi_thread_basic_gpt/49_multi_thread_basic.c
use std::sync::{Arc, Mutex};
use std::thread;

fn withdraw(account_balance: &Arc<Mutex<i32>>) {
    let amount = 500000;
    for _ in 0..amount {
        let mut balance = account_balance.lock().unwrap();
        *balance -= 1;
    }
}

fn main() {
    let account_balance = Arc::new(Mutex::new(1000000));

    let balance_clone = Arc::clone(&account_balance);
    let t1 = thread::spawn(move || {
        withdraw(&balance_clone);
    });

    let balance_clone = Arc::clone(&account_balance);
    let t2 = thread::spawn(move || {
        withdraw(&balance_clone);
    });

    t1.join().unwrap();
    t2.join().unwrap();

    println!("Final account balance: {}", *account_balance.lock().unwrap());
}

File: ga/4omini_result_2024_7/result/multi_thread_basic_cpp_mini/49_multi_thread_basic.rs
use std::sync::Arc;
use std::sync::Mutex;
use std::thread;

fn main() {
    // Initialize account balance with 1,000,000
    let account_balance = Arc::new(Mutex::new(1_000_000));

    // Create a clone of the Arc for each thread
    let balance_clone1 = Arc::clone(&account_balance);
    let balance_clone2 = Arc::clone(&account_balance);

    // Define the withdrawal function
    let withdraw = |balance: Arc<Mutex<i32>>, amount: i32| {
        // Attempt to withdraw the specified amount
        for _ in 0..amount {
            let mut balance_lock = balance.lock().unwrap();
            *balance_lock -= 1;
        }
    };

    // Start two threads to withdraw 500,000 each
    let t1 = thread::spawn(move || withdraw(balance_clone1, 500_000));
    let t2 = thread::spawn(move || withdraw(balance_clone2, 500_000));

    // Wait for both threads to finish
    t1.join().unwrap();
    t2.join().unwrap();

    // Print the final account balance
    let final_balance = account_balance.lock().unwrap();
    println!("Final account balance: {}", *final_balance);
}